pyfr/readers/cgns.py

```
@@ -0,0 +1,423 @@
+        20: ('mixed',),
+    }
+
+    def __init__(self, cgns, base, idx):
+        self._cgns = cgns
+        zone = cgns.zone_read(base, idx)
+
+        # Read Node
+        self.nodepts = self._read_nodepts(zone)
+
+        # Read bc
+        bc = self._read_bc(zone)
+
+        # Read elements
+        ns = cgns.nsections(zone)
+        self.elenodes = elenodes = {}
+        self.pents = pents = {}
+
+        # Construct elenodes and physical entity
+        newp = 0
+        for idx in range(ns):
+            elerng, elenode = self._read_element(zone, idx)
+
+            name = 'fluid'
+            for bcname, bcrng in bc.items():
+                if elerng[0] >= bcrng[0] and elerng[1] <= bcrng[1]:
+                    name = bcname
+
+            if name in pents:
+                pent = pents[name]
+            else:
+                pent = newp
+                pents[name] = pent
+                newp += 1
+
+            elenodes.update({(k, pent): v for k, v in elenode.items()})
+
+    def _read_nodepts(self, zone):
+        # Node read
+        nnode = zone['size'][0]
+        nodepts = np.zeros((3, nnode))
+
+        self._cgns.coord_read(zone, 'CoordinateX', nodepts[0])
+        self._cgns.coord_read(zone, 'CoordinateY', nodepts[1])
+        self._cgns.coord_read(zone, 'CoordinateZ', nodepts[2])
+
+        return nodepts
+
+    def _read_bc(self, zone):
+        nbc = self._cgns.nbocos(zone)
+        bc = {}
+
+        for idx_bc in range(nbc):
+            boco = self._cgns.boco_read(zone, idx_bc)
+            name = boco['name'].lower()
+            bc[name] = boco['range']
+
+        return bc
+
+    def _read_element(self, zone, idx):
+        s = self._cgns.section_read(zone, idx)
+
+        elerng = s['range']
+        conn = np.zeros(s['dim'], dtype=np.int32)
+        self._cgns.elements_read(s, conn)
+
+        try:
+            cgns_type = s['etype']
+            petype = self._cgns_map[cgns_type][0]
+        except KeyError as err:
```

Just let the error throw, no need to catch and reformat.